[CI] CsvFormatReaderTests testMalformedRowErrorAnchorsOnQuoteOffset failing

**Build Scans:**
- [elasticsearch-intake #40020 / part5](https://gradle-enterprise.elastic.co/s/tdegfzyqjrr3k)
- [elasticsearch-pull-request #147750 / part-5](https://gradle-enterprise.elastic.co/s/iujtvmkenqnuo)
- [elasticsearch-pull-request #147751 / part-5](https://gradle-enterprise.elastic.co/s/nlrmffz7ds3rs)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql-datasource-csv:test" --tests "org.elasticsearch.xpack.esql.datasource.csv.CsvFormatReaderTests.testMalformedRowErrorAnchorsOnQuoteOffset" -Dtests.seed=6CE060934025CDF6 -Dtests.locale=ca -Dtests.timezone=Europe/Vaduz -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.datasource.csv.CsvFormatReaderTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testMalformedRowErrorAnchorsOnQuoteOffset'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected offset annotation, got: line -1:-1: CSV parse error at row [1]: Unclosed quoted field in line [1,hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh… (truncated, 828 chars total) …tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt]; row: 1,hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh… (truncated, 828 chars total) …tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt; set error_mode to skip_row (or null_field) in WITH options to skip and warn instead of failing
```

**Issue Reasons:**
- [main] 3 consecutive failures in test testMalformedRowErrorAnchorsOnQuoteOffset
- [main] 1 consecutive failures in step part5
- [main] 2 consecutive failures in step part-5
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 2 consecutive failures in pipeline elasticsearch-pull-request
- [main] 3 failures in test testMalformedRowErrorAnchorsOnQuoteOffset (60.0% fail rate in 5 executions)
- [main] 2 failures in step part-5 (66.7% fail rate in 3 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (66.7% fail rate in 3 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/docs/changelog/149660.yaml
+++ b/docs/changelog/149660.yaml
@@ -1,0 +1,6 @@
+area: ES|QL
+issues:
+ - 149567
+pr: 149660
+summary: Anchor CSV row errors in fused fast path too
+type: bug

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -558,9 +558,6 @@ tests:
 - class: org.elasticsearch.gateway.PersistedClusterStateServiceTests
   method: testLimitsFileCount
   issue: https://github.com/elastic/elasticsearch/issues/149557
-- class: org.elasticsearch.xpack.esql.datasource.csv.CsvFormatReaderTests
-  method: testMalformedRowErrorAnchorsOnQuoteOffset
-  issue: https://github.com/elastic/elasticsearch/issues/149567
 - class: org.elasticsearch.xpack.esql.CsvIT
   method: test {csv-spec:stats.sumWithOverflowRow}
   issue: https://github.com/elastic/elasticsearch/issues/149574

--- a/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/CsvFormatReader.java
+++ b/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/CsvFormatReader.java
@@ -2081,6 +2081,11 @@ public class CsvFormatReader implements SegmentableFormatReader {
             StringBuilder current = new StringBuilder();
             boolean inQuotes = false;
             int bracketDepth = 0;
+            // Remember where the parser entered the unclosed state so error messages can anchor on
+            // the actual fault site instead of head/tail-truncating a long line and hiding it.
+            // Mirrors the offset tracking in splitCommaDelimiterBracketAwareFields.
+            int quoteOpenAt = -1;
+            int bracketOpenAt = -1;
             int fieldIndex = 0;
             boolean fieldHasNonWhitespace = false;
             boolean trailingFieldHasContent = false;
@@ -2132,12 +2137,14 @@ public class CsvFormatReader implements SegmentableFormatReader {
                 } else if (c == quote && fieldHasNonWhitespace == false) {
                     trailingFieldHasContent = true;
                     inQuotes = true;
+                    quoteOpenAt = i;
                     numericValid = false;
                     i++;
                 } else if (c == '[' && fieldHasNonWhitespace == false) {
                     trailingFieldHasContent = true;
                     if (hasMvcBracketClose(line, i)) {
                         bracketDepth = 1;
+                        bracketOpenAt = i;
                     }
                     if (isProjected) current.append(c);
                     numericValid = false;
@@ -2203,10 +2210,14 @@ public class CsvFormatReader implements SegmentableFormatReader {
             }
 
             if (inQuotes) {
-                throw new MalformedRowException("Unclosed quoted field in line [" + CsvErrorMessages.summarize(line) + "]");
+                throw new MalformedRowException(
+                    "Unclosed quoted field in line [" + CsvErrorMessages.summarizeAround(line, quoteOpenAt) + "]"
+                );
             }
             if (bracketDepth > 0) {
-                throw new MalformedRowException("Unclosed bracket cell in line [" + CsvErrorMessages.summarize(line) + "]");
+                throw new MalformedRowException(
+                    "Unclosed bracket cell in line [" + CsvErrorMessages.summarizeAround(line, bracketOpenAt) + "]"
+                );
             }
 
             int totalFields = trailingFieldHasContent ? fieldIndex + 1 : fieldIndex;

--- a/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/CsvFormatReader.java
+++ b/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/CsvFormatReader.java
@@ -1566,12 +1566,10 @@ public class CsvFormatReader implements SegmentableFormatReader {
             }
         }
         if (inQuotes) {
-            throw new MalformedRowException("Unclosed quoted field in line [" + CsvErrorMessages.summarizeAround(line, quoteOpenAt) + "]");
+            throw MalformedRowException.unclosedQuotedField(line, quoteOpenAt);
         }
         if (bracketDepth > 0) {
-            throw new MalformedRowException(
-                "Unclosed bracket cell in line [" + CsvErrorMessages.summarizeAround(line, bracketOpenAt) + "]"
-            );
+            throw MalformedRowException.unclosedBracketCell(line, bracketOpenAt);
         }
         if (current.length() > 0) {
             entries.add(emitField(current));
@@ -2210,14 +2208,10 @@ public class CsvFormatReader implements SegmentableFormatReader {
             }
 
             if (inQuotes) {
-                throw new MalformedRowException(
-                    "Unclosed quoted field in line [" + CsvErrorMessages.summarizeAround(line, quoteOpenAt) + "]"
-                );
+                throw MalformedRowException.unclosedQuotedField(line, quoteOpenAt);
             }
             if (bracketDepth > 0) {
-                throw new MalformedRowException(
-                    "Unclosed bracket cell in line [" + CsvErrorMessages.summarizeAround(line, bracketOpenAt) + "]"
-                );
+                throw MalformedRowException.unclosedBracketCell(line, bracketOpenAt);
             }
 
             int totalFields = trailingFieldHasContent ? fieldIndex + 1 : fieldIndex;

--- a/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/MalformedRowException.java
+++ b/x-pack/plugin/esql-datasource-csv/src/main/java/org/elasticsearch/xpack/esql/datasource/csv/MalformedRowException.java
@@ -27,4 +27,17 @@ final class MalformedRowException extends RuntimeException {
     MalformedRowException(String message, Throwable cause) {
         super(message, cause);
     }
+
+    /**
+     * Shared message template for the structural fault factories below: keeps the two splitter
+     * copies aligned and routes every long-row excerpt through {@link CsvErrorMessages#summarizeAround}
+     * so the printed window is anchored on {@code offset} (the parser's known fault index).
+     */
+    static MalformedRowException unclosedQuotedField(String line, int offset) {
+        return new MalformedRowException("Unclosed quoted field in line [" + CsvErrorMessages.summarizeAround(line, offset) + "]");
+    }
+
+    static MalformedRowException unclosedBracketCell(String line, int offset) {
+        return new MalformedRowException("Unclosed bracket cell in line [" + CsvErrorMessages.summarizeAround(line, offset) + "]");
+    }
 }

